kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,750 @@
+    -- compatible with balancer, which needs this field
+    answers.expire = now() + answers.ttl
+    return answers
+  end
+
+  local processed_answers = {}
+  local cname_answer
+
+  local ttl = self.valid_ttl or 0xffffffff  -- 0xffffffff for maximum TTL value
+
+  for _, answer in ipairs(answers) do
+    answer.name = answer.name:lower()
+
+    if answer.type == TYPE_CNAME then
+      cname_answer = answer   -- use the last one as the real cname
+
+    elseif answer.type == qtype then
+      -- compatible with balancer, see https://github.com/Kong/kong/pull/3088
+      if answer.type == TYPE_AAAA then
+        answer.address = ipv6_bracket(answer.address)
+      elseif answer.type == TYPE_SRV then
+        answer.target = ipv6_bracket(answer.target)
+      end
+
+      -- skip the SRV record pointing to itself,
+      -- see https://github.com/Kong/lua-resty-dns-client/pull/3
+      if not (answer.type == TYPE_SRV and answer.target == qname) then
+        table_insert(processed_answers, answer)
+      end
+    end
+
+    if self.valid_ttl then
+      answer.ttl = self.valid_ttl
+    else
+      ttl = math_min(ttl, answer.ttl)
+    end
+  end
+
+  if #processed_answers == 0 then
+    if not cname_answer then
+      return {
+        errcode = EMPTY_RECORD_EC,
+        errstr  = EMPTY_RECORD_ESTR,
+        ttl     = self.empty_ttl,
+        -- expire = now() + self.empty_ttl,
+      }
+    end
+
+    table_insert(processed_answers, cname_answer)
+  end
+
+  processed_answers.expire = now() + ttl
+  processed_answers.ttl = ttl
+
+  return processed_answers
+end
+
+
+local function resolve_query(self, name, qtype, tries)
+  local key = name .. ":" .. qtype
+  stats_count(self.stats, key, "query")
+
+  local r, err = resolver:new(self.r_opts)
+  if not r then
+    return nil, "failed to instantiate the resolver: " .. err
+  end
+
+  local options = { additional_section = true, qtype = qtype }
+  local answers, err = r:query(name, options)
+  if r.destroy then
+    r:destroy()
+  end
+
+  if not answers then
+    stats_count(self.stats, key, "query_fail_nameserver")
+    return nil, "DNS server error: " .. (err or "unknown")
+  end
+
+  answers = process_answers(self, name, qtype, answers)
+
+  stats_count(self.stats, key, answers.errstr and "query_fail:" .. answers.errstr
+                                               or "query_succ")
+
+  return answers, nil, answers.ttl
+end
+
+
+local function stale_update_task(premature, self, key, name, qtype, short_key, ttl)
+  if premature then
+    return
+  end
+
+  local answers = resolve_query(self, name, qtype, {})
+  if not answers then
+    -- retry update after failure
+    local retry_delay = math_random(STALE_UPDATE_DELAY, STALE_UPDATE_DELAY * 2)
+    ttl = ttl - retry_delay
+    if ttl < 0 then
+      return  -- no need to retry if it exceeds the stale_ttl
+    end
+    local ok, err = timer_at(retry_delay, stale_update_task, self, key, name,
```

Comment:
This may not be as reliable as it could be, should we simply try resolve once in stale mode, and let subsequent stale query trigger the update in the background instead? This seems to give better concurrency control than the retry code triggering a new timer on it's own, which may create self feedback loop.

Reply:
As we talked with @dndx @ADD-SP offline, we decided to update the stale record at an interval, with a maximum spacing of 60s, so here we will insert the stale record back into mlcache shdict with the TTL value of`math.min(left-stale-ttl, 60) s` and let the upper caller to trigger the update task if the stale record expires again.